services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioning.java

```diff
@@ -22,29 +22,122 @@
+     *         </ul>
+     * @throws NullPointerException if any of the parameters except clientContext is {@code null}.
+     */
+    public static Future<CredentialsResult<JsonObject>> provisionIfEnabled(
+            final String tenantId,
+            final String authId,
+            final JsonObject clientContext,
+            final CredentialsManagementService credentialsManagementService,
+            final CredentialsService credentialsService,
+            final DeviceManagementService deviceManagementService,
+            final TenantInformationService tenantInformationService,
+            final Span span) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(authId);
+        Objects.requireNonNull(deviceManagementService);
+        Objects.requireNonNull(credentialsManagementService);
+        Objects.requireNonNull(credentialsService);
+        Objects.requireNonNull(tenantInformationService);
+        Objects.requireNonNull(span);
+
+        return tenantInformationService.getTenant(tenantId, span)
+                .compose(tenantResult -> {
+                    if (tenantResult.isError()) {
+                        return Future.succeededFuture(CredentialsResult.from(tenantResult.getStatus()));
+                    }
+
+                    final TenantObject tenantConfig = tenantResult.getPayload();
+                    return DeviceRegistryUtils
+                            .getCertificateFromClientContext(tenantId, authId, clientContext, span)
+                            .compose(optionalCert -> optionalCert
+                                    .filter(cert -> AutoProvisioning.isEnabled(tenantConfig, cert, span))
+                                    .map(cert -> {
+                                        Tags.ERROR.set(span, Boolean.FALSE); // reset error tag
+                                        return provisionDevice(tenantId, authId, deviceManagementService, credentialsManagementService, cert, span)
+                                                        .compose(r -> {
+                                                            // if error then return error result
+                                                            if (r.isError()) {
+                                                                TracingHelper.logError(span, r.getPayload());
+                                                                return Future.succeededFuture(CredentialsResult
+                                                                        .from(r.getStatus(), new JsonObject()
+                                                                                .put("description", r.getPayload())));
+                                                            }
+                                                            // else return the new credentials
+                                                            return getNewCredentials(credentialsService, tenantId,
```

Comment:
it looks very odd to me that we need to employ the Credentials service to retrieve the newly created credentials here.
Can't *provisionDevice* simply return the new X.509 credentials that it has created?

Reply:
It's a good point :)  We can very well use the new X.509 credentials that it has created. I will update it accordingly.

Reply:
@sophokles73 I have made changes based on the comments and pushed a commit for it. Would you mind taking a look again?